Add reusable blocks

noisysocks opened the pull request:
## ✨ Hello, reusable blocks! ✨ 

![panel-2](https://user-images.githubusercontent.com/612155/31065598-5562562e-a792-11e7-90e3-22c49bca0dd1.gif)

This PR adds all of the frontend code necessary to support viewing, creating, inserting, and editing reusable blocks.

https://github.com/WordPress/gutenberg/issues/1516 describes the feature and https://github.com/WordPress/gutenberg/issues/2081 (https://github.com/WordPress/gutenberg/issues/2081#issuecomment-325231157, in particular) describes the technical approach I'm taking. We're using the backend API that https://github.com/WordPress/gutenberg/pull/2503 added.

## How to test

1. With a Gutenblock selected, click 'Convert to reusable block' in the advanced options sidebar. This creates a new reusable block.
2. Use the inserter to insert a new instance of this reusable block.
3. With a reusable block selected in the editor, click 'Detach from Reusable Block' in the advanced options sidebar. This converts the reusable block back to a regular block.
4. Play with creating, editing and saving different types of reusable blocks.

noisysocks committed:
Add 'core/reusable-block' block type

This lets us parse reusable blocks from the HTML, and not much else.

noisysocks committed:
Stub out Reusable Block UI, implement fetching and local editing

Stubs out some of the editor UI so that we have a base to work from.
Also adds the reducers and effects necessary to support fetching and
local editing.

noisysocks committed:
Implement reusable block persisting

 Adds an effect that responds to the PERSIST_REUSABLE_BLOCK action, and
 some UI for the confirmation dialog that appears when a reusable block
 is first modified.

noisysocks committed:
Implement ATTACH_REUSABLE_BLOCK

Adds an effect which lets us turn reusable blocks into regular blocks.

noisysocks committed:
Implement DETACH_REUSABLE_BLOCK

Adds an effect which lets us turn regular blocks into reusable blocks,
and stubs out some Advanced Options UI which triggers this effect.

noisysocks committed:
Fetch and persist reusable blocks using the WP API Backbone client

Adds the `gutenberg/v1/` endpoint to the WPI API Backbone client so that
we can use `wp.api.models.ReusableBlocks` to fetch and save reusable
blocks instead of our own (ugly) implementation.

noisysocks committed:
Clean up reusable block update actions

Collapse SET_REUSABLE_BLOCK_NAME and UDPATE_REUSABLE_BLOCK_ATTRIBUTES
into one UPDATE_REUSABLE_BLOCK action. This makes things read a little
nicer.

noisysocks committed:
Rename persistReusableBlock -> saveReusableBlock

'persist' is a little verbose, and we already use the term 'save' when
saving the Backbone model.

noisysocks committed:
Prefer 'id' instead of 'ref' when dealing with reusable blocks

This keeps our variables names in sync with how things look in the API,
and disambiguates between the ID stored in the post and the ID stored in
redux.

noisysocks committed:
Rename ATTACH/DETACH actions

Renames ATTACH_REUSABLE_BLOCK -> ATTACH_BLOCK and DETACH_REUSABLE_BLOCK
-> DETACH_BLOCK. This makes it clearer that these actions take block
UIDs, not reusable block IDs.

noisysocks committed:
Add test fixture for core/reusable-block

noisysocks committed:
Write tests for reusable block client API

Write tests for reuasble block actions, selectors, reducer and factory
methods.

noisysocks committed:
Write tests for ATTACH_BLOCK and DETACH_BLOCK effects

noisysocks committed:
Style 'Convert to Reusable Block' button

Introduce a button new inspector control and use it for the 'Convert to
Reusable Block' and 'Detach from Reusable Block' buttons.

noisysocks committed:
Style reusable block dialogs and display them modally

- Introduces a new Modal component, which presents its children modally
- Changes ReusableBlockEdit to present NewReusableBlockDialog and
  SaveConfirmationDialog modally
- Styles the reusable block dialogs to look like the design

noisysocks commented:
I was talking with @nb and the topic of alternative approaches came up. Pasting my thoughts here 😄 

---

The big alternative approach I’ve been mulling over is to enhance the Gutenberg blocks API to the point where reusable blocks could be added as a plain old block type that doesn’t need to touch redux at all.

To do this, the blocks API needs three things:

1) Some way of letting blocks have internal state (think `setState`) that can be initialised from data that is passed into `createBlock`. This could be as simple as having the API let you mark an attribute as being _private_ or _non-serializable_.
2) Some way of letting you provide a higher order matching function instead of a `blocks` array when defining transforms.
3) To pass the complete block in on calls to `transform()` instead of just the block attributes.

If we had these things then reusable blocks could, I think, look _something_ like this:

```js
registerBlockType( 'core/reusable-block', {
    ...
    attributes: {
        ref: { type: 'string' },
        _referencedBlock: { type: 'object' } // the underscore denotes that this is "private" 
    },
    transforms: {
        from: [
            {
                type: 'block',
                match: ( destinationBlock, sourceBlock ) => destinationBlock.name === sourceBlock.attributes._referencedBlock.name,
                transform( sourceBlock ) {
                    return createBlock( sourceBlock.attributes._referencedBlock.name, sourceBlock.attributes._referencedBlock.attributes );
                }
            }
        ],
        to: [
            {
                type: 'block',
                match: () => true,
                transform( sourceBlock ) {
                    return createBlock( 'core/reusable-block', {
                        ref: uuid(),
                        _referencedBlock: sourceBlock,
                    } );
                }
            }
        ],
    }
} );
```

Pros:

- Feels nice having reusable block be _just a block_.
- It's a lot less code.

Cons:
- API changes feel a bit awkward and hacky.
- Reusable blocks aren’t a native part of the editor.

I'm inclined to stick with the redux approach that this PR contains—it's more code, but I feel the explicitness is useful for when this feature evolves into something more advanced.

noisysocks committed:
Add reusable blocks to the inserter

Show reusable blocks that are available in the redux state tree in the
inserter UI.

noisysocks committed:
Add FETCH_REUSABLE_BLOCKS and have the inserter dispatch it when shown

noisysocks committed:
*Always* ask to confirm editing a reusable block

noisysocks commented:
OK, all of the major pieces of reusable blocks are now implemented. I'd love to get some initial feedback on this PR before I start adding tests, cleaning it up, and breaking it up into smaller PRs! ❤️😃 

Here's a lil' demo:

![reusable-blocks](https://user-images.githubusercontent.com/612155/30232123-c1bda48c-94b3-11e7-88e7-c62abfee37f6.gif)

cc. @nb @mtias @westonruter @jasmussen

munirkamal commented:
This looks great @noisysocks 

If I am not missing, the possibility of making multiple blocks or a group of blocks reuseable is also on the roadmap after this?

noisysocks commented:
Good question, @munirkamal! I agree with @jasmussen's sentiments in https://github.com/WordPress/gutenberg/issues/1516#issuecomment-317991224: 

> Now the mockups got me so excited, but what they in fact depict, should probably be for V2 of Gutenberg. And they should probably take advantage of block nesting (see #428). So you could imagine first inserting a "Group" block, and then moving a number of blocks inside that group. Once that is done, you can once again make that single group block into a reusable block.

munirkamal commented:
@noisysocks 

That makes sense, and I also agree with what @jasmussen suggested there.

Keep up the great work folks, Gutenberg is coming along really well.

noisysocks committed:
Give reusable blocks a dashed outline

youknowriad commented:
Awesome work here! 

Before digging into the code, I'd like to think a bit about the "modal" for the reusable block interactions and whether it's a good flow or not.

How does it work now:

 - You try to edit the block, and once the `onChange` is catched, we ask the user if he wants to update the reusable block or detach from the original block
 - If you detach, it's quite obvious, you'll be presented with a "local" block, and the reusable link is "lost"
 - if you updated the reusable block, the question is, when do we ask the user again if he want to continue editing the original block. For now, this is being done when remounting the component (for example moving to text mode and back to edit mode), but this is not ideal. A slightly better approach would be to persist this in local state and never ask until the user refreshes the page.

But still, I find this confusing.

I'd suggest the following flow if possible:

 - Preset the user with a read-only version of the block (calling the `save` function of the reusable block)
 - Show only two buttons in the block toolbar: Edit and Detach
 - Edit triggers the edit mode of the reusable blocks, showing a modal (or in place) the `edit` function of the reusable block and persisting this change when we "submit" it (we need a submit button in this edit mode)
 - Detach behaves like the current approach.

cc @mtias

noisysocks commented:
Thanks for your thoughts @youknowriad!

I agree that the modal is confusing. Something I've been noticing too is that it's _especially_ confusing when there are multiple instances of the same reusable block in the post.

I like the toolbar idea—I'll spend some time experimenting with this approach 😄

jasmussen commented:
Great work all around, and great thoughts too. I think the modals work fine, but I also think Riads suggestion is potentially better.

This is such a nice feature, it might be worth looking at it in two phases. One which features the modal as it is — seems like it might be worth getting this PR merged in and tested as is.

Then in phase two we can take the learnings from V1 and look at the UI improvements that Riad suggested. Unless of course there are technical underpinnings that would need this to change sooner rather than later, in which case I'd defer to Riad (and correct me if I'm wrong).

In both cases, we should do something about the visual appearance of the global block, if we can. Just like how collaborative editing has a colored border, we should have a distinguishing look to a block that is registered as global. It could, for example, have a dashed line instead of the solid line (and the way to do this as we learned from the collab editing stuff is add a CSS class to global blocks and assign the dashed border there). 

I do agree with Riad that if the block appears to be completely like other blocks, editable and all, it's easy to forget that it's global, and the modal then becomes jarring. Perhaps an intermediate fix is to dim add a very subtle gray dimming overlay to the entire block (perhaps instead of, or in addition to the dashed border), change the cursor to a pointer instead of the caret. You'd then be able to click the block to get the "detach" prompt, instead of having this happen onChange.

jasmussen commented:
Robert brought to my attention a problem with the idea suggested above:

> I do agree with Riad that if the block appears to be completely like other blocks, editable and all, it's easy to forget that it's global, and the modal then becomes jarring. Perhaps an intermediate fix is to dim add a very subtle gray dimming overlay to the entire block (perhaps instead of, or in addition to the dashed border), change the cursor to a pointer instead of the caret. You'd then be able to click the block to get the "detach" prompt, instead of having this happen onChange.

In this situation you couldn't ever _select_ a block (to move it or delete it, say), because that would trigger the prompt. 

Which means we're back to either being fine with the onchange prompt for V1 and then making Riads suggested improvements for V2, or doing those changes before merging. I'm fine with either, but would defer to @karmatosed, @mtias or @youknowriad's feelings on this matter.

However you are able to edit a global block, it should still have a distinct style, seems like it's worth trying with a dashed line.

jasmussen commented:
After further discussions with Robert, we mocked up this:

<img width="1043" alt="screen shot 2017-09-28 at 09 06 57" src="https://user-images.githubusercontent.com/1204802/30953479-6feccf06-a42c-11e7-889f-4d78c675c245.png">

That's what happens when you select a reusable block — a panel at the bottom pops out at the bottom. When the block is just selected, it has "Edit" and "Detach" buttons, as well as a name there. When you've clicked "Edit", the name becomes editable, and there's a "Save" button there.

Thoughts? 

Also, is that still a V2, or should that be V1?

noisysocks commented:
> Also, is that still a V2, or should that be V1?

I'd be happy to do it for V1. The two approaches are much of a muchness, code wise. Removing the modal would feel kind of nice since it'd mean one less `<Slot>` in the app.

noisysocks committed:
Implement `render_callback` for reusable blocks

Allow reusable blocks to be rendered by the server for posts and pages
by registering them in PHP and defininig a `render_callback`.

noisysocks committed:
Replace reusable block modals with inline edit panel

Replaces SaveConfirmationDialog, NewReusableBlockDialog, and Modal with
ReusableBlockEditPanel. This panel sits at the bottom of a reusable
block when selected and allows the user to explicitly edit and save the
reusable block. The end result is a less ambiguous and more explicit
UX.

noisysocks commented:
I replaced the modals with the panel that @jasmussen and I discussed. Here's a GIF:

![panel-2](https://user-images.githubusercontent.com/612155/31065598-5562562e-a792-11e7-90e3-22c49bca0dd1.gif)

Check out [this video](https://cldup.com/BgDKj5b0Tx.mov) for a higher quality demonstration.

I'm really into it. Thoughts?

jasmussen commented:
@noisysocks Nice, that looks good! 

Question: are the "Edit" and "Detach" buttons as well as the title field hidden when the block is unselected? If so then 👍 👍 — see also https://github.com/WordPress/gutenberg/blob/master/docs/design.md#block-design-checklist-dos-and-donts-and-examples

Can you make it so the panel itself either has a background (imagine that quick doodle above had a gray background same as the border color), or a border separator above the fields? Just sort of separate content from chrome so to speak. 

Does it work well on mobile? I.e. does it use available space in a good way? Remember you can use flexbox if you need the textfield to resize to available space. Buttons might be localized and grow bigger. 

Given those tweaks I think we're ship shape to go!

mtias commented:
@jasmussen this should be handled the same way in which we'll handle blocks that save to places outside of the post (i.e a site title block). Those blocks, at least in the post editor incarnation, would likely need their own "save" mechanism for clarity, so they should all do a similar thing.

noisysocks commented:
> are the "Edit" and "Detach" buttons as well as the title field hidden when the block is unselected?

Yep!

> Can you make it so the panel itself either has a background

It has a light grey background right now. You can't see it in the low quality GIF, but if you watch the the [video that I linked to](https://cldup.com/BgDKj5b0Tx.mov), you'll spot it 😄 

> Does it work well on mobile?

Yeah, good point. I'll spend some time optimising it for small screens today.

noisysocks commented:
Tweaked the styles a bit to perform better on mobile. Here's how it all looks:

| Desktop  | Mobile |
| ------------- | ------------- |
| ![vagrant local-wp-wp-admin-admin php-page gutenberg post_id 419 3](https://user-images.githubusercontent.com/612155/31110050-f22d47f8-a84f-11e7-88b3-4580042835c0.png) | ![vagrant local-wp-wp-admin-admin php-page gutenberg post_id 419 iphone 6 2](https://user-images.githubusercontent.com/612155/31110038-e702fb8e-a84f-11e7-9973-987a3ba12a2a.png) |
| ![vagrant local-wp-wp-admin-admin php-page gutenberg post_id 419 1](https://user-images.githubusercontent.com/612155/31110104-494b12a4-a850-11e7-9097-3bda9390c1b1.png)   | ![vagrant local-wp-wp-admin-admin php-page gutenberg post_id 419 iphone 6](https://user-images.githubusercontent.com/612155/31110107-503cb13a-a850-11e7-90e1-c7a5fbf8cd88.png)  |
| ![vagrant local-wp-wp-admin-admin php-page gutenberg post_id 419 2](https://user-images.githubusercontent.com/612155/31110109-56df20a4-a850-11e7-9cf8-e356ace1b14d.png)   | ![vagrant local-wp-wp-admin-admin php-page gutenberg post_id 419 iphone 6 1](https://user-images.githubusercontent.com/612155/31110123-7430a10a-a850-11e7-8c3c-1c99f8fc5cbb.png)  |

jasmussen commented:
I like those changes, nice work 👍 👍

@mtias as I'm still trying to grok the full extent of your comments above, can you chime in also? Thanks.

youknowriad commented on blocks/inspector-controls/button-control/index.js:
```diff
@@ -0,0 +1,21 @@
+import './style.scss';
+
+function ButtonControl( { instanceId, label, value, help, ...props } ) {
+	const id = 'inspector-button-control-' + instanceId;
```
Is the instanceId and the id useful in this component? Generally, we're using it to match labels and inputs but there's no input in this component.

noisysocks replied:
Good catch. `<button>` elements [are labelable](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Form_labelable), but I wasn't passing `id` along to the `<Button>` component. I've fixed this in 224eb27f54a397b3c64f3793174fcb9c2e0cd252.

youknowriad commented on blocks/library/reusable-block/index.js:
```diff
@@ -0,0 +1,154 @@
+		// TODO: Can these two actions be combined?
+		// TODO: Think about a loading indicator, success message, failure message, etc.
+		this.props.updateReusableBlock( pickBy( { name, attributes } ) );
+		this.props.saveReusableBlock();
```
Yes, maybe the `saveReusableBlock`action could take the updated name and attributes as arguments and optimistically update the state (similarly to how we do this for the `savePost` effect).

noisysocks replied:
The only problem is that `saveReusableBlock` is also dispatched [when converting a static block into a reusable block](https://github.com/WordPress/gutenberg/pull/2659/files#diff-c0e459b150c84795a6a30e7104d1027fR386). But maybe I could combine `addReusableBlock`, `updateReusableBlock`, **and** `saveReusableBlock` into one action... 🤔

youknowriad commented on blocks/library/reusable-block/index.js:
```diff
@@ -0,0 +1,154 @@
+ * TODO:
+ * These should use selectors and action creators that we gather from context.
+ * OR this entire component should move to `editor`, I can't decide.
+ */
```
We're in the process of merging the two modules (blocks and editor) which will resolve this issue #2795

noisysocks replied:
Nice! I was dreading how awkward using `context` here would be! 😅 

I'll leave the code as is until #2795 is merged.

youknowriad replied:
#2795 is closed. It was too disruptive for now but the Reusable block shows precisely that our current separation is not great @aduth

youknowriad commented on blocks/library/reusable-block/index.js:
```diff
@@ -0,0 +1,154 @@
+
+registerBlockType( 'core/reusable-block', {
+	title: __( 'Reusable Block' ),
+	category: null,
```
What does a `null` category mean? If I understand correctly, this hides the block from the inserter. this is probably just a side effect of how the inserter is built. Should we be more explicit and add a "private" flag instead?

noisysocks replied:
Yeah, that's right. I made it so that `category` can be marked `null` so as to have it not appear in the inserter. 

A more explicit flag is a good idea. I've introduced this in ddc491b32e7e8bcc269caf220401f7005713f9ad.

youknowriad commented on blocks/library/reusable-block/index.php:
```diff
@@ -0,0 +1,54 @@
+		return $block_type->render( $block_attributes, $block_content );
+	} else {
+		return $block_content;
+	}
```
I wonder if we can factorize some code between here and the `do_blocks` function into a `render_block( block )` function

noisysocks replied:
Great catch! Cleaned up in ccf3024e1f317f918463e956f967e949e0514b51 🌈

youknowriad commented on blocks/library/reusable-block/index.php:
```diff
@@ -0,0 +1,54 @@
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
+	if ( $block_type ) {
+		return $block_type->render( $block_attributes, $block_content );
+	} else {
```
Minor: the `else` is useless here.

noisysocks replied:
👌 b1562b90c7cd5cb00eb63539568ec1c23bd2ca46

youknowriad commented on editor/effects.js:
```diff
@@ -273,4 +284,106 @@ export default {
+		dispatch( addReusableBlocks( reusableBlock ) );
+		dispatch( saveReusableBlock( reusableBlock.id ) );
+		dispatch( replaceBlocks( [ oldBlock.uid ], [ newBlock ] ) );
+	},
```
I'm confused, are these two actions reversed (attache/detach). Attach should create a reusable block and detach should create a regular block (as I understand it at least)

noisysocks replied:
Yeah, I wrote this all thinking that attach would create a regular block since you're taking a block that is _unattached_ from the post (i.e., it's saved separately from the post) and _attaching_ it to a post.

In my view, the whole attach terminology is confusing. I was thinking of making the actions `MAKE_BLOCK_REUSABLE` and `MAKE_BLOCK_STATIC`. Thoughts?

youknowriad replied:
> I was thinking of avoiding the issue altogether by making the actions MAKE_BLOCK_REUSABLE and MAKE_BLOCK_STATIC. Thoughts?

sounds good to me 👍

noisysocks replied:
Fixed in f1368ba4 📈

youknowriad commented on editor/effects.js:
```diff
@@ -273,4 +284,106 @@ export default {
 		return effects;
 	},
+	// TODO: FETCH_REUSABLE_BLOCKS and FETCH_REUSABLE_BLOCK can probably be combined.
+	FETCH_REUSABLE_BLOCKS( action, store ) {
```
can't we replace all those custom fetching actions with the `withApiData`? Maybe not because we want these in the Redux state? in which case, maybe there's something we can do here. A wrapper to `withApiData` dealing with the state or something? @aduth any thoughts on this?

noisysocks replied:
Like you say, we'd need to change `withApiData` to store the models it fetches in Redux. It could be a cool abstraction. I feel it's best to punt on it for now, though.

gziolo replied:
This is exactly what I'm looking for today :) It would be nice to be able to have an option to use the same API without a component, e.g. inside Redux middleware.

noisysocks committed:
Give the `<button>` in `<ButtonControl>` an `id`

This makes the `<label>` in the control behave correctly. According to
the spec, `<button>` is a labelable element:
https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Form_labelable

noisysocks committed:
Use an explicit `isPrivate` flag to hide block types from the inserter

We don't wish to show the 'core/reusable-block' block type in the
inserter. We were doing this by taking advantage of the fact that the
inserter would ignore block types with a null category. This change
makes this behaviour more explicit.

noisysocks committed:
Remove useless `} else {`

noisysocks committed:
DRY up `do_blocks` and `gutenberg_render_block_core_reusable_block`

Refactor the common logic in `do_blocks` and
`gutenberg_render_block_core_reusable_block` into a new
`gutenberg_render_block` function.

noisysocks committed:
Combine FETCH_REUSABLE_BLOCK and FETCH_REUSABLE_BLOCKS

Simplify fetching logic a little by DRYing up these two mostly-similar effects.

noisysocks committed:
Remove ADD_REUSABLE_BLOCKS action, add ability to track saving

Collapses the ADD_REUSABLE_BLOCKS action into UPDATE_REUSABLE_BLOCK and
FETCH_REUSABLE_BLOCKS_SUCCESS. Also makes it so that the Redux store
keeps track of whether or not a reusable block is being saved or has had
an error while saving.

noisysocks committed:
Add reusable block 'saving' and 'save failure' UI

Adds a spinner that indicates whether or not a reusable block is saving,
and some UI that allows the user to retry the save if it fails.

noisysocks commented:
I cleaned some of the redux actions up and added these really simple progress and failure states to the UI:

<img width="640" alt="saving" src="https://user-images.githubusercontent.com/612155/31205824-a0e94a56-a9bf-11e7-89ed-d6786319cfcb.png">
<img width="640" alt="failed" src="https://user-images.githubusercontent.com/612155/31205830-a9e515ea-a9bf-11e7-9ca7-27ae25511538.png">

noisysocks committed:
Rename {ATTACH,DETACH}_BLOCK to MAKE_BLOCK_{STATIC,REUSABLE}

Avoid ambiguity around what 'attach' and 'detach' means by avoiding the
its use in code alltogether.

noisysocks commented:
I've moved some things around, cleaned up some of the redux actions, and added these really simple progress and failure states to the UI:

<img width="640" alt="saving" src="https://user-images.githubusercontent.com/612155/31205824-a0e94a56-a9bf-11e7-89ed-d6786319cfcb.png">
<img width="640" alt="failed" src="https://user-images.githubusercontent.com/612155/31205830-a9e515ea-a9bf-11e7-9ca7-27ae25511538.png">

I think this feature is looking pretty good. I'm going to start breaking it up into smaller PRs that are easier to review and merge 💃

noisysocks committed:
Improve tests for actions, selectors and effects

Slightly clean up the tests for reusable block actions and selectors,
and change our effects tests to reduce our dependency on mocking
selectors by taking advantage of the fact that MAKE_BLOCK_STATIC and
MAKE_BLOCK_REUSABLE are relatively pure functions.

noisysocks committed:
Add unit tests for FETCH_REUSABLE_BLOCKS and SAVE_REUSABLE_BLOCK

youknowriad commented:
@noisysocks Nice job so far, I'm fine with merging this PR directly without splitting, Could you rebase it, now that the REST API is merged? It'll make reviewing easier :)

mtias commented:
> I've moved some things around, cleaned up some of the redux actions, and added these really simple progress and failure states to the UI.

I wonder if these are better as global notices (the post published one), as it provides a consistent place for their display. I can see the proximity being useful, but also can see how it could be missed if you scroll, look elsewhere. cc @jasmussen @karmatosed

jasmussen commented:
I think I agree with Matías, it seems like notices would be a good place for this. Although I'm conflicted because it totally makes sense if there's a spinner in the area that's saving, it makes sense to see message in context of that. Tammie, final word?

mtias commented:
I think the "save" button could have the same treatment we use for the "publish" button with the stripes. That signals something is happening. But the confirmation then happens via global notices.

aduth commented on editor/inserter/menu.js:
```diff
@@ -283,7 +306,7 @@ export class InserterMenu extends Component {
 				key={ block.name }
 				className="editor-inserter__block"
-				onClick={ this.selectBlock( block.name ) }
+				onClick={ this.selectBlock( block.name, block.attributes ) }
```
I find it a bit odd at a glance that in arguments to "selecting" a block we pass attributes. What does attributes have to do with a selection? Seems like this is a convenience for something we could be looking up later in the logic to map that selection into a block creation?

noisysocks replied:
Yeah, that's a good point. I'll play with making `selectBlock` pass up an _enum-ish_ object (e.g. `{ name: 'core/paragraph' }` or `{ name: 'core/reusable-block', ref: '6c08c662-3825-4f9b-9512-ff8a2d209a6b' }`) instead of a string.

noisysocks replied:
I think the right thing to do here is to refactor `InserterMenu` so that it no longer calls `getCategories`, `getBlockTypes`, `getBlocks`, etc. Instead, `InserterMenu` would be a "dumb" component that renders a list of `items` that `Inserter` passes it and calls `item.onSelect()` when one of these items is selected.

Some code to illustrate what I mean:

```js
const staticItems = getBlockTypes()
	.filter( ( blockType ) => ! blockType.isPrivate )
	.map( ( blockType ) => ( {
		category: blockType.category,
		icon: blockType.icon,
		title: blockType.title,
		isDisabled: blockType.useOnce && find( blocks, ( { name } ) => blockType.name === name ),
		isRecent: find( recentlyUsedBlocks, ( { name } ) => blockType.name === name ),
		onSelect() {
			onInsertBlock( blockType.name, {}, insertionPoint );
			onClose();
		},
	} ) );

const reusableItems = /* omitted for brevity */;

return (
	<InserterMenu
		categories={ getCategories() }
		items={ [ ...staticItems, ...reusableItems ] } />
);
```

My fear though that this is quite a lot to change in this already large PR—I'd prefer to ticket this and come back to it later.

aduth commented on editor/layout/index.js:
```diff
@@ -48,6 +49,7 @@ function Layout( { mode, isSidebarOpened, notices, ...props } ) {
 				<MetaBoxes />
 			</div>
 			{ isSidebarOpened && <Sidebar /> }
+			<Slot name="Editor.Modal" fillChildProps={ { isSidebarOpened } } />
```
You noted a comment here: https://github.com/WordPress/gutenberg/pull/2659#issuecomment-332750653

>I'd be happy to do it for V1. The two approaches are much of a muchness, code wise. Removing the modal would feel kind of nice since it'd mean one less <Slot> in the app.

Should we be removing this?

noisysocks replied:
Oops! Fixed in 883c7835

aduth commented on lib/class-wp-rest-reusable-blocks-controller.php:
```diff
@@ -0,0 +1,353 @@
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_Reusable_Blocks_Controller extends WP_REST_Controller {
```
A reusable block is saved as just a post, yeah? Could we be leveraging more of the existing posts endpoint? The endpoint itself? `extends WP_REST_Posts_Controller`?

aduth commented on lib/client-assets.php:
```diff
@@ -791,9 +791,13 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
+	wp_add_inline_script(
+		'wp-editor',
+		(
+			'jQuery.when( wp.api.init(), wp.api.init( { versionString: \'gutenberg/v1/\' } ) ).done( function() {'
```
Do we need to initialize twice like this?

`jQuery` is not an explicit dependency of `wp-editor` and should be added as one if we're going to call on it. Currently it works likely because it's incidentally a dependency of another script on the page, but we can't rely on it to exist this way.

noisysocks replied:
`wp_add_inline_script` doesn't seem to support specifying dependencies and so I've added a manual check for jQuery in 3d1b66c1.

> Do we need to initialize twice like this?

I'm... not sure.

I couldn't find any documentation about this stuff. [The source](https://github.com/WP-API/client-js/blob/master/js/load.js#L367) seems to imply that `init()` is to be called once for every API root and version string. In our case, we have two version strings: `/wp/v2` and `/gutenberg/v1`. I'd love for someone who knows more about the WP API client to chime in 😅

aduth reviewed (commented):
In taking this for a test-drive prior to looking at code, it works very well and feels quite cool. Nice work 👍 

To code, my two overall concerns are:

- How much reusable blocks needs to be integrated into global state of the application, vs. isolated to the individual reusable block implementation
- The forced distinction we're creating between reusable blocks and posts, which are effectively the same, but we're implementing many things (REST API endpoint, UUID as queryable slug) as if they weren't.

noisysocks committed:
Clean up <ButtonControl>

Remove unnecessary {}s, split up long line, and collapse SCSS selectors
into one line.

noisysocks committed:
Clarify how we're using pickBy

Add a comment which clarifies this weird way we're using _.pickBy.

noisysocks committed:
Call `blockType.save`, not `blockType.create`

Fix this typo. There's no such thing as `blockType.create`!

noisysocks commented on blocks/library/reusable-block/index.php:
```diff
@@ -0,0 +1,44 @@
+ */
+function gutenberg_render_block_core_reusable_block( $attributes ) {
+	$reusable_blocks = get_posts( array(
+		'post_type' => 'gb_reusable_block',
```
I wanted this to be `gutenberg_reusable_block` but post types can't have a name longer than 20 characters. Maybe `gutenberg_block` would be better?

atimmer replied:
Given everything is a block in gutenberg, would `gutenberg_reusable` not be enough? Or `gutenberg_block` given only reusable blocks are persisted in a different post type.

noisysocks committed:
Use convertBlockTo{Static,Reusable} instead of makeBlock{Static,Reusable}

It's a clearer action name since 'convert' is a less ambigious verb.

noisysocks committed:
Remove unnecessary `<Slot>`

noisysocks commented on blocks/inspector-controls/button-control/index.js:
```diff
@@ -0,0 +1,21 @@
+function ButtonControl( { instanceId, label, value, help, ...props } ) {
+	const id = 'inspector-button-control-' + instanceId;
+	return (
+		<BaseControl id={ id } label={ label } help={ help } className={ 'blocks-button-control' }>
```
:ok_hand: Fixed in 134114d42a40fbd6e15310623d3b7458474def40.

noisysocks commented on blocks/inspector-controls/button-control/index.js:
```diff
@@ -0,0 +1,21 @@
+	const id = 'inspector-button-control-' + instanceId;
+	return (
+		<BaseControl id={ id } label={ label } help={ help } className={ 'blocks-button-control' }>
+			<Button { ...props } id={ id } isLarge className={ 'blocks-button-control__button' }>{ value }</Button>
```
:ok_hand: Fixed in 134114d42a40fbd6e15310623d3b7458474def40.

noisysocks commented on blocks/inspector-controls/button-control/style.scss:
```diff
@@ -0,0 +1,5 @@
+.blocks-button-control {
+	.blocks-button-control__button {
```
:ok_hand: Fixed in 134114d42a40fbd6e15310623d3b7458474def40.

noisysocks commented on blocks/library/reusable-block/index.js:
```diff
@@ -0,0 +1,160 @@
+
+	updateReusableBlock() {
+		const { name, attributes } = this.state;
+		this.props.updateReusableBlock( pickBy( { name, attributes } ) );
```
👍 Fixed in 93217b32.

noisysocks commented on blocks/library/reusable-block/index.js:
```diff
@@ -0,0 +1,160 @@
+		}
+
+		const blockType = getBlockType( reusableBlock.type );
+		const BlockEdit = blockType.edit || blockType.create;
```
Oops! Good catch, this should be `blockType.edit || blockType.save`. Fixed in bc4a930e.

noisysocks commented on editor/actions.js:
```diff
@@ -321,3 +321,74 @@ export const createSuccessNotice = partial( createNotice, 'success' );
+}
+
+/**
+ * Returns an action object used to convert a reusable block into a static
```
👍 Fixed in d8a6ac4a.

noisysocks committed:
Ensure jQuery is loaded before using it in an inline script

noisysocks commented on blocks/library/reusable-block/index.php:
```diff
@@ -0,0 +1,44 @@
+function gutenberg_render_block_core_reusable_block( $attributes ) {
+	$reusable_blocks = get_posts( array(
+		'post_type' => 'gb_reusable_block',
+		'name' => $attributes['ref'],
```
From https://github.com/WordPress/gutenberg/issues/2081:

> A block would be identified by a UUID which would get saved in post_name. Using a UUID instead of auto-incremented post ID would eliminate the requirement that a block post be created up-front to reserve its ID. The auto-incremented post ID would be for internal use only and not exposed in the API.

noisysocks commented on blocks/library/reusable-block/index.js:
```diff
@@ -0,0 +1,160 @@
+
+	componentDidMount() {
+		if ( ! this.props.reusableBlock ) {
+			this.props.fetchReusableBlock();
```
It *would* be nice to simplify this all with some abstraction. I'm not sure that `withAPIData` is the right abstraction, though.

The major problem with using `withAPIData` is that our 'Detach from Reusable Block' functionality needs to have the attributes of the reusable block that it is detaching. Currently, it gets this by [accessing the store](https://github.com/WordPress/gutenberg/pull/2659/files#diff-c0e459b150c84795a6a30e7104d1027fR349). If we moved the fetched reusable block to inside the `edit` component's local state, then we would not be able to access this from outside of the component.

Also: right now, if you have two instances of the same reusable block in the post, editing one will immediately update the other. We get this for free because the UI is derived from what's in the store. It's a small thing, but kind of nice.

noisysocks committed:
Use animation and notices to indicate the reusable block save state

- Makes the 'Save' button animate when a reusable block is saving
- Makes a 'Reusable block updated' notice appear when the save is
  successful
- Makes a 'Reusable block update failed' notice appear when the save
  fails

noisysocks commented:
> I think the "save" button could have the same treatment we use for the "publish" button with the stripes. That signals something is happening. But the confirmation then happens via global notices.

I've implemented this in ec61310a7ce17982aa199cea5e53b4bb294a54a6 😄

noisysocks commented:
Thanks for all the feedback, everyone! 🙏

I'm closing this and splitting up the work into three smaller PRs. The first is right here: https://github.com/WordPress/gutenberg/pull/3017
